### ISSUE TITLE
Handle user updating the width of their window dynamically

### ISSUE THREAD

**Issue by bitspittle** [varabyte/kotter#34]

**Comment by bitspittle** [comment]
This might not be possible. I can't seem to find ways to either

1) stop the terminal from autowrapping my lines
2) ask the terminal its width in a consistent manner
3) ask the terminal to tell me when it added a newline

There may be a way and it's probably worth leaving this bug open, but not necessarily as a 1.0 milestone. My experimentation with other dynamic CLIs is they suffer a similar fate.

**Comment by bitspittle** [comment]
Actually, the "$COLUMNS" variable in Linux seems to work. With seeing if this is standard-enough behavior that we can use it consistently.

**Comment by bitspittle** [comment]
It's still quite tricky to get the width. It's a shame either jline doesn't support this OR it does but I can't find it. (jline's terminal.getSize() doesn't update dynamically as you resize the window)

See also: https://stackoverflow.com/questions/1286461/can-i-find-the-console-width-with-java/18883172#18883172

**Comment by bitspittle** [comment]
Leaving the bug open because we still don't handle dynamic window resizing, and I'd like to think that's possible someday. But for now, at least we take initial window size into account, which is better than nothing, so I'm going to remove the 1.0 milestone from this bug.

**Comment by bitspittle** [comment]
I'm going to close this as dynamic window resizing just doesn't seem like something that consoles are good at.

**Comment by bitspittle** [comment]
Now that I have a bunch of K/N code, I'm actually going to reopen this and check one more time to make sure.

Really, what I need is an event that the column changed (or maybe I can poll this every frame).

I'm pretty sure I tried polling when using Jline, but I should try again now that I've got K/N solutions.

**Comment by 239** [comment]
> It's a shame either jline doesn't support this OR it does but I can't find it.

There is a fairly recent example for terminal size handling: [TerminalSizeExample.java](https://github.com/jline/jline3/blob/master/demo/src/main/java/org/jline/demo/examples/TerminalSizeExample.java)

It is also mentioned in the troubleshooting guide: [Terminal Size Issues](https://jline.org/versions/4.0/docs/troubleshooting#terminal-size-issues)

**Comment by bitspittle** [comment]
Very interesting, thanks! I'm very interested in trying this out.

I will probably hold off post 1.3.0 before working on this. It might go smoothly or it might be a nightmare if there are any knockdown surprising affects!

**Comment by bitspittle** [comment]
I'm happy to say that this is now available for testing in the snapshot (`1.3.1-SNAPSHOT`).

As of this moment, I only support the virtual terminal and the JVM system terminal. I will be looking into K/N soon.

There are no docs yet but the feature is fairly simple -- just use `width` and `height` vars in your `section` and then resize the terminal.

```kotlin
section {
   textLine("$width x $height")
   textLine("Press Q to quit")
}.runUntilKeysPressed(Keys.Q)
```

After K/N is done, I will write docs, do my best to test that I didn't break anything major across the three major platforms, and then I'll 
add some sort of "responsive" example, maybe a table that changes from 3 columns to 2 to 1 as you shrink the window. I'm not sure yet.

P.S. At this point the VIrtual Terminal has been significantly refactored from the ground up, so there may be bugs there as well. Most of my last week has been spent trying to test every possible nasty edge case I could think of, and at this point all the examples do seem to run well. But if anyone tries anything and sees that it's busted for them, let me know!

**Comment by bitspittle** [comment]
K/N should be done. It wasn't too bad. I have not tested windows yet but the code to add in theory was trivial, so I hope it will work. I will test it tomorrow (it's getting late here!)

**Comment by 239** [comment]
It works really well and makes testing dynamic layouts so much easier 👍 Thank you!

> At this point the VIrtual Terminal has been significantly refactored from the ground up, so there may be bugs there as well.

I guess this is one of them (only in VT), symbols at the bottom are halved:

<img width="320" height="230" alt="Image" src="https://github.com/user-attachments/assets/7e795461-5e08-4bc6-9581-bd51dc021fdb" />

**Comment by bitspittle** [comment]
Thanks so much for checking the feature out! After assuming it was impossible so many years ago, it ended up being pretty trivial to add, and the result feels so clean. (Just use $width and/or $height and you've opted into refreshes)

Oh wow how am I missing render width?? Thanks for the report, I'll take a look at that.

**Comment by bitspittle** [comment]
Hmmm I can't reproduce. I'm using the code from [this comment](https://github.com/varabyte/kotter/issues/154#issuecomment-4900919331). Maybe you have an updated project you can share?

My screen:
<img width="926" height="49" alt="Image" src="https://github.com/user-attachments/assets/593420c2-fe40-49d9-90b2-0c87ff1335b2" />

This is a new issue yeah? You weren't seeing it before?

(Side note: I'll try on Windows later, maybe it's a font issue?)

**Comment by 239** [comment]
Sorry for the sparse info and delay, now I can try it out more extensively!

> Hmmm I can't reproduce. I'm using the code from [this comment](https://github.com/varabyte/kotter/issues/154#issuecomment-4900919331).

Strange, it is still the old code using the latest snapshot and your modified code has the same issue. System terminals show full symbols 🫠

https://github.com/user-attachments/assets/f9da7f1b-44ff-4f78-aa74-db9405645481

> This is a new issue yeah? You weren't seeing it before?

Yes, new one. I just tested the previous snapshot and it shows regular symbols!

**Comment by 239** [comment]
It is related to `invert()` somehow, full symbols without inverting 🤔

**Comment by bitspittle** [comment]
> It is related to `invert()` somehow, full symbols without inverting 🤔

That's very interesting. Thank you -- I'll dig into that lead.

**Comment by bitspittle** [comment]
@239 still can't repro.

Do you see it with this program?

```kotlin
const val chars = "♔♕♖♗♘♙♚♛♜♝♞♟"

fun main() = session {
    var shouldInvert by liveVarOf(false)
    section {
        if (shouldInvert) invert()
        textLine(chars)
    }.runUntilKeyPressed(Keys.Q) {
        onKeyPressed {
            if (key == Keys.Space) {
                shouldInvert = !shouldInvert
            }
        }
    }
}
```

(Press SPACE to toggle invert, Q to quit)

**Comment by 239** [comment]
Yes, but it depends on the terminal and it's **not** only the VT:

**Ghostty** (full chars)

https://github.com/user-attachments/assets/a49e5e17-fc35-4225-8150-63bc88ce8618

**Konsole** (last char only)

https://github.com/user-attachments/assets/bea6dfa3-d651-4113-8d2b-041b1f9c292d

**Virtual**

https://github.com/user-attachments/assets/c7aa9394-e356-4713-be7b-56a99d097d1e

I also changed fonts but that seems to have no effect...

**Comment by 239** [comment]
**Windows Terminal**

<img width="150" height="200" alt="Image" src="https://github.com/user-attachments/assets/b292e412-47d6-44fb-b4ca-8c7816efe0c6" />

**Comment by bitspittle** [comment]
Wow that's so weird.

Can you check this program in `Konsole` and `Windows Terminal` when targeting Kotter 1.3.0? I'm guessing it will _still_ be broken in that case, but the Virtual Terminal will work.

I'm going to pull out a Windows machine today and see if I can repro over there. The code seems to work fine on my machine. I suspect it's going to be a windows font issue -- like, I'm treating it as characters with a render width of 1, but on windows specifically they should have a render width of 2. My terminal uses the default system mono font, but the chess pieces probably aren't in it.

And maybe 1.3.0 virtual terminal worked because the Swing text area renderer did something smart based on the active _actual_ font being used to render the character, not sure.

Anyway my top goal for today is to try to reproduce myself! If I can do that I think I can dig into a fix much faster.

**Comment by bitspittle** [comment]
P.S. super useful screen captures, thank you so much!

**Comment by bitspittle** [comment]
OK I can reproduce and can even see several ways to solve it.

I'm going to close this issue as fixed and open a new one!

**Comment by 239** [comment]
Noticed your update too late...
